strawberry/permission.py

```diff
@@ -1,21 +1,184 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Any, Awaitable, Optional, Union
+import abc
+from inspect import iscoroutinefunction
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Awaitable,
+    Dict,
+    List,
+    Optional,
+    Type,
+    Union,
+)
+
+from strawberry.exceptions import StrawberryGraphQLError
+from strawberry.exceptions.permission_fail_silently_requires_optional import (
+    PermissionFailSilentlyRequiresOptionalError,
+)
+from strawberry.extensions import FieldExtension
+from strawberry.schema_directive import Location, StrawberrySchemaDirective
+from strawberry.type import StrawberryList, StrawberryOptional
+from strawberry.utils.await_maybe import await_maybe
+from strawberry.utils.cached_property import cached_property
 
 if TYPE_CHECKING:
-    from strawberry.types.info import Info
+    from graphql import GraphQLError, GraphQLErrorExtensions
+
+    from strawberry.extensions.field_extension import (
+        AsyncExtensionResolver,
+        SyncExtensionResolver,
+    )
+    from strawberry.field import StrawberryField
+    from strawberry.types import Info
 
 
-class BasePermission:
+class BasePermission(abc.ABC):
     """
     Base class for creating permissions
     """
 
     message: Optional[str] = None
 
+    error_extensions: Optional[GraphQLErrorExtensions] = None
+
+    error_class: Type[GraphQLError] = StrawberryGraphQLError
+
+    _schema_directive: Optional[object] = None
+
+    @abc.abstractmethod
     def has_permission(
         self, source: Any, info: Info, **kwargs: Any
     ) -> Union[bool, Awaitable[bool]]:
         raise NotImplementedError(
             "Permission classes should override has_permission method"
         )
+
+    def handle_no_permission(self) -> None:
```

Comment:
shall we call this `handle_unauthorized`? or maybe `on_unauthorized` to be consistent with extensions?

Reply:
sounds good 👍